RFCs/FS-1135-random-functions-for-collections.md

```diff
@@ -0,0 +1,206 @@
+
+[ArgumentException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentoutofrangeexception) should be raised if N is greater than collection length or is negative
+
+[ArgumentException](https://learn.microsoft.com/en-us/dotnet/api/system.argumentexception) should be raised if collection is empty
+
+Example:
+```fsharp
+let allPlayers = [ "Alice"; "Bob"; "Charlie"; "Dave" ]
+let round1Order = allPlayers |> List.randomSample 3 // ["Charlie", "Dave", "Alice"]
+```
+
+# Drawbacks
+
+System.Random interface has added some new methods in [.NET 8](https://learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-8#methods-for-working-with-randomness), where naming is a bit different. More new methods can eventually be added in future .NET versions.
+
+# Alternatives
+
+Use online snippets, or provide a nuget package.
+
+# Compatibility
+
+Please address all necessary compatibility questions:
+
+* Is this a breaking change? **No, unless user defined those extensions themselves**
+* What happens when previous versions of the F# compiler encounter this design addition as source code? **Should be fine**
+* What happens when previous versions of the F# compiler encounter this design addition in compiled binaries? **Should be fine**
+* If this is a change or extension to FSharp.Core, what happens when previous versions of the F# compiler encounter this construct? **Will work as usual**
+
+# Pragmatics
+
+## Diagnostics
+
+Please list the reasonable expectations for diagnostics for misuse of this feature. **I don't see a way to misuse it**
+
+## Tooling
+
+Please list the reasonable expectations for tooling for this feature, including any of these:
+
+* Debugging
+    * Breakpoints/stepping
+    * Expression evaluator
+    * Data displays for locals and hover tips
+* Auto-complete
+* Tooltips
+* Navigation and Go To Definition
+* Colorization
+* Brace/parenthesis matching
+
+**Should work just like for other collections functions**
+
+## Performance
+
+Please list any notable concerns for impact on the performance of compilation and/or generated code
+
+* For existing code **Existing code should not be affected**
+* For the new features **Performance should be respected when implementing this feature, since it can be used in performance-sensitive scenaris**
+
+## Scaling
+
+Algorithmic complexity of the new features should be O(n) or less. 
+
+## Culture-aware formatting/parsing
+
+N/A
+
+# Unresolved questions
+
+It's unclear, if more sophisticated overloads should be added:
+ - Weights parameter for randomChoices function
+ - Counts parameter for randomSample function
+
+~~In .NET 6 and higher [Random.Shared](https://learn.microsoft.com/en-us/dotnet/api/system.random.shared) is available, but as soon as F# Core only targets standard, it can't use it. Is targeting higher .NET possible (not just for this feature, maybe some others need it)?
```

Comment:
No, not possible (not yet, at least). It would mean maintaining multiple versions of FSharp.Core and we are trying really, really hard to prevent that.

I think you an drop this as _unresolved question_. We will just cherry-pick from the existing implementation, as suggested above.

Reply:
Fixed